Gordon360/Services/HousingService.cs

```diff
@@ -537,5 +560,215 @@ public bool ChangeApplicationDateSubmitted(int applicationID)
+                    if (student != null)
+                    {
+                        yearList.Add(student.Class);
+                    }
+                }
+            }
+            var newMaxYear = new Year
+            {
+                ApplicationID = applicantion_id,
+                Year1 = yearList.Max()
+            }; ;
+            await _context.Year.AddAsync(newMaxYear);
+            _context.SaveChanges();
+        }
+
+        /// <summary>
+        /// update the information of preferred halls
+        /// </summary>
+        /// <param name="username"> The username for the user who complete the aplication form </param>
+        /// <param name="applicantion_id"> The ID of this application </param>
+        /// <param name="hallList"> A list of the preferred halls </param>
+        public async Task UpdateHallAsync(string username, string applicantion_id, string[] hallList)
+        {
+            var GordonID = _context.ACCOUNT.FirstOrDefault(a => a.AD_Username == username)?.gordon_id;
+            if (GordonID == null)
+            {
+                throw new ResourceNotFoundException() { ExceptionMessage = "The applicant could not be found" };
+            }
+            var existHall = _context.PreferredHall.Where(x => x.ApplicationID == applicantion_id);
+            foreach (PreferredHall h in existHall)
+            {
+                _context.PreferredHall.Remove(h);
+            }
+            var rank = 1;
+            foreach (string h in hallList)
+            {
+                if (h != "")
+                {
+                    var newHall = new PreferredHall
+                    {
+                        ApplicationID = applicantion_id,
+                        Rank = rank,
+                        HallName = h
+                    }; ;
+                    await _context.PreferredHall.AddAsync(newHall);
+                    rank++;
+                }
+            }
+            _context.SaveChanges();
+        }
+
+        /// <summary>
+        /// update the information of preferred halls
+        /// </summary>
+        /// <param name="username"> The username for the user who complete the aplication form </param>
+        /// <param name="applicantion_id"> The ID of this application </param>
+        /// <param name="preferenceList"> A list of the preference </param>
+        public async Task UpdatePreferenceAsync(string username, string applicantion_id, string[] preferenceList)
+        {
+            var GordonID = _context.ACCOUNT.FirstOrDefault(a => a.AD_Username == username)?.gordon_id;
+            if (GordonID == null)
+            {
+                throw new ResourceNotFoundException() { ExceptionMessage = "The applicant could not be found" };
+            }
+            var existPreference = _context.Preference.Where(x => x.ApplicationID == applicantion_id);
+            foreach (Preference p in existPreference)
+            {
+                _context.Preference.Remove(p);
```

Comment:
Why remove and then re-add preferences? That has a lot of room for error. Instead, you should get the preference row in question, update it, then commit it back to the database.